types/evidence.go

```diff
@@ -342,3 +355,581 @@ func (evl EvidenceList) Has(evidence Evidence) bool {
+			case 1:
+				i++
+				continue OUTER_LOOP
+			case -1:
+				j++
+			}
+		}
+	}
+
+	return evList
+}
+
+func (ev ConflictingHeadersEvidence) Height() int64 { return ev.H1.Height }
+
+// XXX: this is not the time of equivocation
+func (ev ConflictingHeadersEvidence) Time() time.Time { return ev.H1.Time }
+
+func (ev ConflictingHeadersEvidence) Address() []byte {
+	panic("use ConflictingHeadersEvidence#Split to split evidence into individual pieces")
+}
+
+func (ev ConflictingHeadersEvidence) Bytes() []byte {
+	return cdcEncode(ev)
+}
+
+func (ev ConflictingHeadersEvidence) Hash() []byte {
+	bz := make([]byte, tmhash.Size*2)
+	copy(bz[:tmhash.Size-1], ev.H1.Hash().Bytes())
+	copy(bz[tmhash.Size:], ev.H2.Hash().Bytes())
+	return tmhash.Sum(bz)
+}
+
+func (ev ConflictingHeadersEvidence) Verify(chainID string, _ crypto.PubKey) error {
+	panic("use ConflictingHeadersEvidence#VerifyComposite to verify composite evidence")
+}
+
+// VerifyComposite verifies that both headers belong to the same chain, same
+// height and signed by 1/3+ of validators at height H1.Height == H2.Height.
+func (ev ConflictingHeadersEvidence) VerifyComposite(committedHeader *Header, valSet *ValidatorSet) error {
+	var alternativeHeader *SignedHeader
+	switch {
+	case bytes.Equal(committedHeader.Hash(), ev.H1.Hash()):
+		alternativeHeader = ev.H2
+	case bytes.Equal(committedHeader.Hash(), ev.H2.Hash()):
+		alternativeHeader = ev.H1
+	default:
+		return errors.New("none of the headers are committed from this node's perspective")
+	}
+
+	// ChainID must be the same
+	if committedHeader.ChainID != alternativeHeader.ChainID {
+		return errors.New("alt header is from a different chain")
+	}
+
+	// Height must be the same
+	if committedHeader.Height != alternativeHeader.Height {
+		return errors.New("alt header is from a different height")
+	}
+
+	// Limit the number of signatures to avoid DoS attacks where a header
+	// contains too many signatures.
+	//
+	// Validator set size               = 100 [node]
+	// Max validator set size = 100 * 2 = 200 [fork?]
+	maxNumValidators := valSet.Size() * 2
+	if len(alternativeHeader.Commit.Signatures) > maxNumValidators {
+		return errors.Errorf("alt commit contains too many signatures: %d, expected no more than %d",
+			len(alternativeHeader.Commit.Signatures),
+			maxNumValidators)
+	}
+
+	// Header must be signed by at least 1/3+ of voting power of currently
+	// trusted validator set.
+	if err := valSet.VerifyCommitTrusting(alternativeHeader.ChainID, alternativeHeader.Commit.BlockID, alternativeHeader.Height,
+		alternativeHeader.Commit, tmmath.Fraction{Numerator: 1, Denominator: 3}); err != nil {
+		return errors.Wrap(err, "alt header does not have 1/3+ of voting power of our validator set")
+	}
+
+	return nil
+}
+
+func (ev ConflictingHeadersEvidence) Equal(ev2 Evidence) bool {
+	switch ev2.(type) {
+	case ConflictingHeadersEvidence:
+		e2 := ev2.(ConflictingHeadersEvidence)
+		return bytes.Equal(ev.H1.Hash(), e2.H1.Hash())
```

Comment:
Do you want to compare both header hashes or do you think one is sufficient? If a chain was forked into three, then you could have different conflicting headers being equal - not sure if this is a problem or not

Reply:
No, you're right. We should compare both headers